vql/server/splunk.go

```diff
@@ -0,0 +1,244 @@
+	SkipVerify bool                `vfilter:"optional,field=skip_verify,doc=Skip SSL verification(default: False)."`
+
+	WaitTime int64 `vfilter:"optional,field=wait_time,doc=Batch splunk upload this long (2 sec)."`
+}
+
+type _SplunkPlugin struct{}
+
+func (self _SplunkPlugin) Call(ctx context.Context,
+	scope *vfilter.Scope,
+	args *ordereddict.Dict) <-chan vfilter.Row {
+	output_chan := make(chan vfilter.Row)
+	scope.Log("set up splunk")
+
+	go func() {
+		defer close(output_chan)
+
+		err := vql_subsystem.CheckAccess(scope, acls.COLLECT_SERVER)
+		if err != nil {
+			return
+		}
+
+		arg := _SplunkPluginArgs{}
+		err = vfilter.ExtractArgs(scope, args, &arg)
+		if err != nil {
+			return
+		}
+
+		if arg.Threads == 0 {
+			arg.Threads = 1
+		}
+
+		if arg.ChunkSize == 0 {
+			arg.ChunkSize = 1000
+		}
+
+		if arg.WaitTime == 0 {
+			arg.WaitTime = 2
+		}
+
+		if len(arg.Sourcetype) == 0 {
+			arg.Sourcetype = "vql"
+		}
+
+		if len(arg.Source) == 0 {
+			arg.Source = "velociraptor"
+		}
+
+		wg := sync.WaitGroup{}
+		row_chan := arg.Query.Eval(ctx, scope)
+		for i := 0; i < int(arg.Threads); i++ {
+			wg.Add(1)
+
+			// Start an uploader on a thread.
+			go _upload_rows(ctx, scope, output_chan,
+				row_chan, &wg, &arg)
+		}
+
+		wg.Wait()
+	}()
+	return output_chan
+}
+
+// Copy rows from row_chan to a local buffer and push it up to splunk.
+func _upload_rows(
+	ctx context.Context,
+	scope *vfilter.Scope, output_chan chan vfilter.Row,
+	row_chan <-chan vfilter.Row,
+	wg *sync.WaitGroup,
+	arg *_SplunkPluginArgs) {
+	defer wg.Done()
+
+	var buf []*ordereddict.Dict
+
+	client := splunk.NewClient(
+		&http.Client{
+			Timeout: time.Second * 20,
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{InsecureSkipVerify: arg.SkipVerify},
+			},
+		}, // Optional HTTP Client objects
+		arg.URL,
+		arg.Token,
+		arg.Source,
+		arg.Sourcetype,
+		arg.Index,
+	)
+
+	wait_time := time.Duration(arg.WaitTime) * time.Second
+	next_send_time := time.After(wait_time)
+
+	// Flush any remaining rows
+	defer send_to_splunk(scope, output_chan, client, &buf, arg)
+
+	// Batch sending to splunk: Either
+	// when we get to chuncksize or wait
+	// time whichever comes first.
+	for {
+		select {
+		case row, ok := <-row_chan:
+			if !ok {
+				return
+			}
+
+			//
+			err := _append_row_to_buffer(ctx, scope, row, &buf, arg)
+			if err != nil {
+				continue
+			}
+
+		case <-next_send_time:
+
+			send_to_splunk(scope, output_chan,
+				client, &buf, arg)
+
+			next_send_time = time.After(wait_time)
+		}
+	}
+}
+
+func _append_row_to_buffer(
+	ctx context.Context,
+	scope *vfilter.Scope,
+	row vfilter.Row, buf *[]*ordereddict.Dict,
+	arg *_SplunkPluginArgs) error {
+
+	_buf := *buf
+	row_dict := vfilter.RowToDict(ctx, scope, row)
+
+	// if ClientID exists and "Host" isn't a field, use the ClientID field
+	clientid, client_idpres := row_dict.Get("ClientId")
```

Comment:
Maybe move to below the `if !host_pres`?   I guess in the common case we should have the VQL emit the host field?

Reply:
I thought about it - and I think you are right. Its best to make the plugin as stupid as possible and let the user control the host field. If the user wants the artifacts to have the host field as the client ID its simply `ClientID as Host`. I don't want to override anyones VQL with the plugin since its supposed to simply stream